descriptor: check whitespace in keys within fragments

brunoerg opened the pull request:
Currently, we successfully parse descriptors which contains spaces in the beginning or end of the public/private key within a fragment (e.g. `pk( KEY)`, `pk(KEY )` or `pk( KEY )`). I have noticed that one of the reasons is that the `DecodeBase58` function simply ignore these whitespaces. 

This PR changes the `ParsePubkeyInner ` to reject pubkeys that contain a whitespace at the beginning and/or at the end. We will only check the whitespace in some RPCs (e.g. `importdescriptors`), but an already imported descriptor won't be affected by this check, especially because we store descriptors from `ToString`.

For context: https://github.com/brunoerg/bitcoinfuzz/issues/72

brunoerg committed:
descriptor: check whitespace in ParsePubkeyInner

Due to Base58, keys with whitespace at the beginning or
at the end are successfully parsed. This commit adds a
check into `ParsePubkeyInner` to verify whether if the
first or last char of the key is a space.

brunoerg committed:
test: descriptor: check whitespace into keys

furszy reviewed (commented):
As you are forbidding something that was previously allowed, this breaks compatibility with earlier versions. Wallets containing any key with a whitespace in the descriptor will fail to parse during the load procedure and shut down abruptly.
Will either need a descriptor upgrade procedure or a compatibility + warning window.

brunoerg commented:
> As you are forbidding something that was previously allowed, this breaks compatibility with earlier versions. Wallets containing any key with a whitespace in the descriptor will fail to parse during the load procedure and shut down abruptly. Will either need a descriptor upgrade procedure or a compatibility + warning window.

Good point. Moved it to draft to work on these things.

brunoerg committed:
test: `getdescriptorinfo`/`importdescriptors` with whitespace in pubkeys

brunoerg committed:
docs: add release notes for 31603

brunoerg commented:
Ready for review!

Thanks, @furszy for the comment. I've changed the approched and described it on description.

darosior commented:
> Wallets containing any key with a whitespace in the descriptor will fail to parse during the load procedure and shut down abruptly.

As far as i remember we only ever store in the wallet descriptor produced by `ToString()`, which would mean no such wallet can exist. This would merely be an interface break for RPC's which parse descriptors, given the lack of usecase of using a space inside a `pk()` expression this would be pretty minor.

furszy commented:
> > Wallets containing any key with a whitespace in the descriptor will fail to parse during the load procedure and shut down abruptly.
> 
> As far as i remember we only ever store in the wallet descriptor produced by `ToString()`, which would mean no such wallet can exist. This would merely be an interface break for RPC's which parse descriptors, given the lack of usecase of using a space inside a `pk()` expression this would be pretty minor.

Was thinking about that a few days ago and started checking it, but discovered another bug along the way and wanted to fix that before commenting here again.
Storing the `ToString()` output means we store the descriptor with a different checksum, which could be confusing for users calling `listdescriptors`. However, it's the lesser of the evils, so the compatibility window might not be necessary. A test for this would be useful, verifying that we are actually storing the descriptor without whitespaces, which results in a different checksum from the one provided by the user.

Note about the test: this can't be done for taproot descriptors that do not contain all key/script paths key material (this is the bug I discovered, [wip branch](https://github.com/furszy/bitcoin-core/commits/2025_descriptors_fix_listdescriptors_for_taproot_internal_pk)).

brunoerg commented:
> As far as i remember we only ever store in the wallet descriptor produced by ToString(), which would mean no such wallet can exist. This would merely be an interface break for RPC's which parse descriptors, given the lack of usecase of using a space inside a pk() expression this would be pretty minor.

It means I can go back to the previous approach (simpler one) and only change the `ParsePubkeyInner`, there is no need to control it in the descriptor `Parse` anymore.

brunoerg committed:
test: fix descriptors in `ismine_tests`

Some descriptors contain whitespace in public keys
within fragments. This fixes it.

furszy commented on src/script/descriptor.cpp:
```diff
@@ -1515,6 +1515,9 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
     if (str.size() == 0) {
         error = "No key provided";
         return {};
+    } else if (IsSpace(str[0]) || IsSpace(str.back())) {
+        error = strprintf("Pubkey '%s' is invalid due to whitespace", str);
+        return {};
```
nit: Because there is a `return` statement in the if block that is above, you could remove the `else` and use `str.front()` instead of `str[0]`.

brunoerg replied:
Done.

brunoerg commented on doc/release-notes-31603.md:
```diff
@@ -0,0 +1,6 @@
+Updated RPCs
+--------
+
+- Any RPC which one of the parameters are descriptors will throw an error
```
Done

furszy commented on src/script/descriptor.cpp:
```diff
@@ -1516,6 +1516,10 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
         error = "No key provided";
         return {};
     }
+    if (IsSpace(str.front()) || IsSpace(str.back())) {
+        error = strprintf("Pubkey '%s' is invalid due to whitespace", str);
+        return {};
+    }
```
This could also be a WIP formatted private key right?
Should use the general term "key" here. E.g. 
```c++
strprintf(key '%s is invalid due to an extra whitespace at %s of the string', key_str, (IsSpace(str.front())?"beginning" : "end"));`
```

brunoerg replied:
Yes, it could be a WIF private key. I will change it to use "key".

brunoerg replied:
Done.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/31603#discussion_r1920553269 (thanks, @furszy)

brunoerg commented:
Forgot to fix a test, force-pushed doing it. Now CI is happy!

rkrux reviewed (commented):
Somewhere either in the PR title or description, it should be mentioned that this works for descriptors containing private keys as well as observed in the tests and in the `ParsePubKeyInner` code: https://github.com/bitcoin/bitcoin/blob/28dec6c5f8bd35ef4e6cb8d7aa3f21b6879acf98/src/script/descriptor.cpp#L1540-L1550

sipa reviewed (commented):
utACK bdc6ac10caf92901657b4d95bf2c11cccc8ac9ea

It's unfortunate that this in theory breaks RPC importing for (very limited) positions where spaces were allowed before, but it makes sense for consistency. Given that wallets always store descriptors in normalized form, I don't believe this will result in wallet loading breaking.

brunoerg commented on doc/release-notes-31603.md:
```diff
@@ -0,0 +1,6 @@
+--------
+
+- Any RPC in which one of the parameters are descriptors will throw an error
+if the provided descriptor contains an whitespace at the beginning or end of
```
Done.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/31603#discussion_r1959636120, https://github.com/bitcoin/bitcoin/pull/31603#discussion_r1959639872 and https://github.com/bitcoin/bitcoin/pull/31603#discussion_r1959643968.

It adds 2 more test cases and fix a comment. Thanks, @rkrux.

darosior commented on src/script/descriptor.cpp:
```diff
@@ -1509,6 +1509,10 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
         error = "No key provided";
         return {};
     }
+    if (IsSpace(str.front()) || IsSpace(str.back())) {
```
*(comment about the commit, not this line)*

nit: in commit e91a9cf9d2ca50ee31ed36a975d2e01faa666c92 you state:
> Due to Base58, pubkeys with whitespace at the beginning
end/or at the end are successfully parsed.

But pubkeys are never base58 encoded.

brunoerg replied:
I meant due to `DecodeExt{Pub}Key` which calls `DecodeBase58Check`.

rkrux commented:
One way to ensure code structure correctness and get rid of the brittleness noticed here (https://github.com/bitcoin/bitcoin/pull/31603#discussion_r1959636120) is to incorporate the following change. In a cursory attempt, I have this that has minimal diff that passes the tests. LMK how does this sound.

```cpp
void CheckUnparsable(const std::string& prv, const std::string& pub, const std::string& expected_error, bool check_prv_error_only = false)
{
    FlatSigningProvider keys_priv, keys_pub;
    std::string error;
    auto parse_priv = Parse(prv, keys_priv, error);
    BOOST_CHECK_MESSAGE(parse_priv.empty(), prv);

    if (check_prv_error_only) {
       BOOST_CHECK_EQUAL(error, expected_error);
       return;
    }
    
    auto parse_pub = Parse(pub, keys_pub, error);
    BOOST_CHECK_MESSAGE(parse_pub.empty(), pub);
    BOOST_CHECK_EQUAL(error, expected_error);
}
```

The function call site with the private key descriptor passed in the correct function argument. 
```cpp
CheckUnparsable("multi(1, L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1,5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjss)", "", "Multi: Key ' L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1' is invalid due to whitespace", /* check_prv_error_only*/ true);
CheckUnparsable("pk(L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1 )", "", "pk(): Key 'L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1 ' is invalid due to whitespace", /* check_prv_error_only*/ true);
CheckUnparsable("pk( L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1 )", "", "pk(): Key ' L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1 ' is invalid due to whitespace", /* check_prv_error_only*/ true);
```

darosior commented on src/script/descriptor.cpp:
```diff
@@ -1509,6 +1509,10 @@ std::vector<std::unique_ptr<PubkeyProvider>> ParsePubkeyInner(uint32_t key_exp_i
         error = "No key provided";
         return {};
     }
+    if (IsSpace(str.front()) || IsSpace(str.back())) {
```
Interestingly this line is correct only because we check `IsHex` before `ParseHex` below, as otherwise we would need to check for any space character across the string since `ParseHex` would skip those even if they happen in the middle.

Checking for a space character at any position would also improve the error message for instance when parsing `pk(03 a34b...)`. But no need to retouch.

brunoerg replied:
Ok.

darosior reviewed (approved):
ACK 35d5adbcd58e04991bb4651d2dd97af6e186d1f9

darosior commented:
> is to incorporate the following change.

Improving the descriptor unit test utilities can be done in a followup. It would be nice to get the user-facing fix in first.

brunoerg commented:
> > is to incorporate the following change.
> 
> Improving the descriptor unit test utilities can be done in a followup. It would be nice to get the user-facing fix in first.

Yes, any other minor stuff I will leave for a follow-up.

rkrux commented:
\

> > > is to incorporate the following change.
> > 
> > 
> > Improving the descriptor unit test utilities can be done in a followup. It would be nice to get the user-facing fix in first.
> 
> Yes, any other minor/test stuff I will leave for a follow-up.

A separate PR is fine. I raised this point because this PR introduces a way of using `CheckUnparsable` in a way that leads to some difficulty in reading the test.

rkrux reviewed (commented):
Review @ 35d5adb

I reviewed range-diff
```
git range-diff bdc6ac1...35d5adb
```

Newer changes are fixing adding new assertions in the functional tests and improving comments.

I find the new error more detailed and helpful compared to the previous one.
```
➜  src git:(master) ✗ bitcoinclireg getdescriptorinfo "multi(1, L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1,5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjss)"
error code: -5
error message:
Multi: key ' L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1' is not valid
➜  src git:(2025-01-descriptor-pk) ✗ bitcoinclireg getdescriptorinfo "multi(1, L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1,5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjss)"
error code: -5
error message:
Multi: Key ' L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1' is invalid due to whitespace
```

Re: https://github.com/bitcoin/bitcoin/pull/31603#pullrequestreview-2623448041
I do need to spend some time in understanding how storing descs in normalised form helps in the wallet loading not breaking, will verify and then approve.

rkrux commented:
> Currently, we successfully parse descriptors which contains spaces in the beginning or end of the public/private key within a fragment 

On master, I am not able to successfully import a descriptor or get the descriptor info in case a space is passed in the desc. Is this statement incorrect or is there any other flow I don't know about?

<details>
<summary>
`getdescriptorinfo` and `importdescriptors` with both public key and private keys on master
</summary>

```
➜  src git:(master) ✗ bitcoinclireg getdescriptorinfo "multi(1, L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1,5KYZdUEo39z3FPrtuX2QbbwGnNP5zTd7yyr2SC1j299sBCnWjss)"
error code: -5
error message:
Multi: key ' L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1' is not valid

➜  src git:(master) ✗ bitcoinclireg getdescriptorinfo "pkh( 02df706d76d09602b3b34a845ae611c2d77449a6a67fc911053d6c39728b697ddf)#atgzwtd5"
error code: -5
error message:
pkh(): key ' 02df706d76d09602b3b34a845ae611c2d77449a6a67fc911053d6c39728b697ddf' is not valid


➜  src git:(master) ✗ bitcoinclireg importdescriptors "[{\"desc\":\"pkh( 02df706d76d09602b3b34a845ae611c2d77449a6a67fc911053d6c39728b697ddf)#atgzwtd5\",\"timestamp\":\"now\"}]"
[
  {
    "success": false,
    "error": {
      "code": -5,
      "message": "pkh(): key ' 02df706d76d09602b3b34a845ae611c2d77449a6a67fc911053d6c39728b697ddf' is not valid"
    }
  }
]

➜  src git:(master) ✗ bitcoinclireg importdescriptors "[{\"desc\":\"pkh( L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1)#e58j46cd\",\"timestamp\":\"now\"}]"
[
  {
    "success": false,
    "error": {
      "code": -5,
      "message": "pkh(): key ' L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1' is not valid"
    }
  }
]
```
</details>

brunoerg commented:
> > Currently, we successfully parse descriptors which contains spaces in the beginning or end of the public/private key within a fragment
> 
> On master, I am not able to successfully import a descriptor or get the descriptor info in case a space is passed in the desc. Is this statement incorrect or is there any other flow I don't know about?
> 
> `getdescriptorinfo` and `importdescriptors` with both public key and private keys on master

@rkrux Some examples I just tried:
```
./build/src/bitcoin-cli getdescriptorinfo "sh(multi(2, 5KASdZKHt7MBqD6tDHTyTsLMfX6E8TXfygKanpPsi5T5xxV7NWR, L1uf6enQ3adcTFdTsiU9ESFpifcYudGS7dh9NhMxUJZZoGvMwh4A))"
{
  "descriptor": "sh(multi(2,04ca1af27a61a0bb32210056e14cf4a9af7429ceebacab06fe8366a020c758395aeb0560daeba81a705c3ae537ca37f3a254bc9a95de1eab897c7ae17c7ccee425,03cacf75b5d27844a3181275c523baf8069707f53f894967d3b549219d0b5880b7))#q6gqlkrs",
  "checksum": "gt4hkzw7",
  "isrange": false,
  "issolvable": true,
  "hasprivatekeys": true
}
```

```
./build/src/bitcoin-cli getdescriptorinfo "pk(L4rK1yDtCWekvXuE6oXD9jCYfFNV2cWRpVuPLBcCU2z8TrisoyY1 )"
{
  "descriptor": "pk(03a34b99f22c790c4e36b2b3c2c35a36db06226e41c692fc82b8b56ac1c540c5bd)#9pcxlpvx",
  "checksum": "gzu3vsx5",
  "isrange": false,
  "issolvable": true,
  "hasprivatekeys": true
}
```

rkrux reviewed (commented):
Re: https://github.com/bitcoin/bitcoin/pull/31603#issuecomment-2675329745

My CLI command called the regtest node in which those mainnet valid private keys failed to parse, which is correct behaviour. I tried getting that descriptor info in the mainnet and was able to do so in master, which is erroneous behaviour, and I see the newly added error in this branch when I try to get the descriptor info or import it. 

I noticed that I could import a space containing descriptor in master but when I listed the descriptor it naturally contained the key without space and the checksum of the descriptor changed as well, rightfully so. But this behaviour is inconsistent with user expectations. Failing the descriptor RPCs when a descriptor with space is provided seems to be the right approach.

<details>
<summary>
Examples in master - one each with pub and WIF priv
</summary>

```
➜  bitcoin git:(master) ✗ bitcoinclireg getdescriptorinfo "wpkh( tpubDCw3vMU3FSDX7txrkYJ4WMyS8urS4EtwZ3hXGKBg29WcuN2BQ7i4VeCtG16BdgwgLZdjajL36PWj5xMSYQEHooJAdHZhyVogLUfumecqNf6/0/*)#rwhpmnvw"
{
  "descriptor": "wpkh(tpubDCw3vMU3FSDX7txrkYJ4WMyS8urS4EtwZ3hXGKBg29WcuN2BQ7i4VeCtG16BdgwgLZdjajL36PWj5xMSYQEHooJAdHZhyVogLUfumecqNf6/0/*)#ud465w0w",
  "checksum": "rwhpmnvw",
  "isrange": true,
  "issolvable": true,
  "hasprivatekeys": false
}
➜  bitcoin git:(master) ✗ bitcoinclireg getdescriptorinfo "pkh(cPksSFRHeu9Aah9qYxFkvPqJBgJ4uEhRpjeGwFEH8BS9gSevd3QT )#hwxvtlaa"
{
  "descriptor": "pkh(020c540089aaabec12258e4ea068b1e95a12a87857c8bd10e1e297c6ad11b5609e)#0sd2vs6c",
  "checksum": "hwxvtlaa",
  "isrange": false,
  "issolvable": true,
  "hasprivatekeys": true
}
```

</details>

I will test with few more keys.

rkrux commented on src/script/descriptor.cpp:
> Due to Base58, pubkeys with whitespace at the beginning
end/or at the end are successfully parsed. This commit
add a parameter into `ParsePubkeyInner` to control whether
it should return an error if the first or last char of the
pubkey is a space.

Comment about the commit message: The latest code doesn't contain any parameter in `ParsePubkeyInner` that controls if the error should be returned.

Also, `pubkeys with whitespace` should be changed to `keys with whitespace` as private keys are also affected as well and pubkeys barring extended keys are not parsed successfully currently in master.

brunoerg replied:
The commit message was outdated, I just pushed fixing it, thanks for noticing!

rkrux reviewed (commented):
Ok, I tested the behaviour on master with more keys and found that the RPC was successful for space containing WIF private keys, xprvs, ans xpubs, but a space containing pubkey returned an error related to invalidity as seen in the below examples. 

For a space containing pubkey, in the code all the `IsHex()`, `DecodeSecret()`, `DecodeExtKey()`, and `DecodeExtPubKey()` functions fail because of the space, the 33/34 bytes size check, presence of 0, presence of 0 respectively. Thus, the `key '%s' is not valid` error is thrown.

Concept ACK 35d5adbcd58e04991bb4651d2dd97af6e186d1f9

I will ack once the commit message for 71d7901613988a43e6faee5deeb8c004c2bf5059 is updated as it is outdated.

<details>
<summary>
Examples tested on master
</summary>

```
➜  bitcoin git:(master) ✗ bitcoinclireg getdescriptorinfo "pkh( tprv8ZgxMBicQKsPeNLUGrbv3b7qhUk1LQJZAGMuk9gVuKh9sd4BWGp1eMsehUni6qGb8bjkdwBxCbgNGdh2bYGACK5C5dRTaif9KBKGVnSezxV/1)#qwgdxsx8"
{
  "descriptor": "pkh(tpubD6NzVbkrYhZ4XqNGAWGWSzmxGWFwVjVTjZxh2fioKbVYi7Jx8fdbprVWsdW7mHwqjchBVas8TLZG4Xwuz4RKU4iaCqiCvoSkFCzQptqk5Y1/1)#4uzxxr48",
  "checksum": "qwgdxsx8",
  "isrange": false,
  "issolvable": true,
  "hasprivatekeys": true
}
➜  bitcoin git:(master) ✗ bitcoinclireg getdescriptorinfo "pkh( tpubD6NzVbkrYhZ4XqNGAWGWSzmxGWFwVjVTjZxh2fioKbVYi7Jx8fdbprVWsdW7mHwqjchBVas8TLZG4Xwuz4RKU4iaCqiCvoSkFCzQptqk5Y1/1)#vwk8d6fa"
{
  "descriptor": "pkh(tpubD6NzVbkrYhZ4XqNGAWGWSzmxGWFwVjVTjZxh2fioKbVYi7Jx8fdbprVWsdW7mHwqjchBVas8TLZG4Xwuz4RKU4iaCqiCvoSkFCzQptqk5Y1/1)#4uzxxr48",
  "checksum": "vwk8d6fa",
  "isrange": false,
  "issolvable": true,
  "hasprivatekeys": false
}
➜  bitcoin git:(master) ✗ bitcoinclireg getdescriptorinfo "pkh( 92MC1SSqgEqXtAQBtirfWsBFsamrY9hGq5S9n4XdRz3Wq9iS7Js)#qffggmax"
{
  "descriptor": "pkh(04a4f476c536dffa2454145e16e106e0511bfd842cc512e5f1ca761184301db1a84de2cc560cec7ddb309ac94f52d74717bd9595fa1c1fbb8aec37cb9fed551e50)#4nn9zvsj",
  "checksum": "qffggmax",
  "isrange": false,
  "issolvable": true,
  "hasprivatekeys": true
}
➜  bitcoin git:(master) ✗ bitcoinclireg getdescriptorinfo "pkh( 04a4f476c536dffa2454145e16e106e0511bfd842cc512e5f1ca761184301db1a84de2cc560cec7ddb309ac94f52d74717bd9595fa1c1fbb8aec37cb9fed551e50)#p9pgdcsv"
error code: -5
error message:
pkh(): key ' 04a4f476c536dffa2454145e16e106e0511bfd842cc512e5f1ca761184301db1a84de2cc560cec7ddb309ac94f52d74717bd9595fa1c1fbb8aec37cb9fed551e50' is not valid
```

</details>

brunoerg commented on doc/release-notes-31603.md:
```diff
@@ -0,0 +1,6 @@
+--------
+
+- Any RPC in which one of the parameters are descriptors will throw an error
+if the provided descriptor contains an whitespace at the beginning or the end
```
Done.

brunoerg commented:
Force-pushed addressing https://github.com/bitcoin/bitcoin/pull/31603#discussion_r1966518694 and https://github.com/bitcoin/bitcoin/pull/31603#discussion_r1967529542. Fixed a outdated commit message.

rkrux reviewed (approved):
tACK 21e9d39a3725cd6107b742f0cb97f65b3640201b

Thanks for addressing the feedback.

brunoerg commented:
CI failure is unrelated.

brunoerg commented:
friendly ping: @sipa @darosior

darosior reviewed (approved):
re-ACK 21e9d39a3725cd6107b742f0cb97f65b3640201b

sipa reviewed (commented):
utACK 21e9d39a3725cd6107b742f0cb97f65b3640201b

brunoerg commented:
rfm?

ryanofsky commented:
> rfm?

Agree this looks ready and I could merge it, but it's assigned to @achow101 and she has the most expertise here and didn't comment on the PR at all yet, so I'd be inclined to wait. But feel free to ping me if anyone thinks this should be merged sooner.
